src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDataTable.java

```diff
@@ -207,92 +217,87 @@ private void internalAddrow(JDBCType jdbcType, Object val, Object[] rowValues,
-                // DataTypes.SHORT_VARTYPE_MAX_BYTES
-
-                if (null == val)
-                    rowValues[pair.getKey()] = null;
-                // java.sql.Date, java.sql.Time and java.sql.Timestamp are subclass of java.util.Date
-                else if (val instanceof java.util.Date)
-                    rowValues[pair.getKey()] = val.toString();
-                else if (val instanceof microsoft.sql.DateTimeOffset)
-                    rowValues[pair.getKey()] = val.toString();
-                else if (val instanceof OffsetDateTime)
-                    rowValues[pair.getKey()] = val.toString();
-                else if (val instanceof OffsetTime)
-                    rowValues[pair.getKey()] = val.toString();
-                else
+                    rowValues[pair.getKey()] = bd;
+                    break;
+
+                case DOUBLE:
+                    rowValues[pair.getKey()] = (val instanceof Double) ? (double) val
+                                                                       : Double.parseDouble(val.toString());
+                    break;
+
+                case FLOAT:
+                case REAL:
+                    rowValues[pair.getKey()] = (val instanceof Float) ? (float) val : Float.parseFloat(val.toString());
+                    break;
+
+                case TIMESTAMP_WITH_TIMEZONE:
+                case TIME_WITH_TIMEZONE:
+                case DATE:
+                case TIME:
+                case TIMESTAMP:
+                case DATETIMEOFFSET:
+                case DATETIME:
+                case SMALLDATETIME:
+                    // Sending temporal types as string. Error from database is thrown if parsing fails
+                    // no need to send precision for temporal types, string literal will never exceed
+                    // DataTypes.SHORT_VARTYPE_MAX_BYTES
+
+                    // java.sql.Date, java.sql.Time and java.sql.Timestamp are subclass of java.util.Date
+                    if (val instanceof java.util.Date || val instanceof microsoft.sql.DateTimeOffset
+                            || val instanceof OffsetDateTime || val instanceof OffsetTime)
+                        rowValues[pair.getKey()] = val.toString();
+                    else
+                        rowValues[pair.getKey()] = (String) val;
+                    break;
+
+                case BINARY:
+                case VARBINARY:
+                case LONGVARBINARY:
+                    nValueLen = ((byte[]) val).length;
+
+                    if (nValueLen > currentColumnMetadata.precision) {
+                        currentColumnMetadata.precision = nValueLen;
+                        columnMetadata.put(pair.getKey(), currentColumnMetadata);
+                    }
+                    rowValues[pair.getKey()] = (byte[]) val;
+                    break;
+
+                case CHAR:
+                    if (val instanceof UUID)
```

Comment:
Why is this special conversion for UUID only present on CHAR and not the other string data types?  Out of scope for this review, just seems a little bit odd.